sno/working_copy/sqlserver_adapter.py

```diff
@@ -0,0 +1,179 @@
+from sno import crs_util
+from sno.schema import Schema, ColumnSchema
+
+from sqlalchemy.sql.compiler import IdentifierPreparer
+from sqlalchemy.dialects.mssql.base import MSDialect
+
+
+_PREPARER = IdentifierPreparer(MSDialect())
+
+
+def quote(ident):
+    return _PREPARER.quote(ident)
+
+
+V2_TYPE_TO_MS_TYPE = {
+    "boolean": "bit",
+    "blob": "varbinary",
+    "date": "date",
+    "float": {0: "real", 32: "real", 64: "float"},
+    "geometry": "geography",
+    "integer": {
+        0: "int",
+        8: "tinyint",
+        16: "smallint",
+        32: "int",
+        64: "bigint",
+    },
+    "interval": "nvarchar",
+    "numeric": "numeric",
+    "text": "nvarchar",
+    "time": "time",
+    "timestamp": "datetimeoffset",
+}
+
+MS_TYPE_TO_V2_TYPE = {
+    "bit": "boolean",
+    "tinyint": ("integer", 8),
+    "smallint": ("integer", 16),
+    "int": ("integer", 32),
+    "bigint": ("integer", 64),
+    "real": ("float", 32),
+    "float": ("float", 64),
+    "date": "date",
+    "datetime": "timestamp",
+    "datetime2": "timestamp",
+    "datetimeoffset": "timestamp",
+    "decimal": "numeric",
+    "geography": "geometry",
+    "geometry": "geometry",
+    "interval": "interval",
+    "numeric": "numeric",
+    "nvarchar": "text",
+    "text": "text",
+    "time": "time",
+    "varchar": "text",
+    "varbinary": "blob",
+}
+
+# Types that can't be roundtripped perfectly in SQL Server, and what they end up as.
+APPROXIMATED_TYPES = {"interval": "text"}
```

Comment:
Seems to support interval types? https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/interval-data-types?view=sql-server-ver15

Reply:
Good catch - added "interval" -> "interval", removed this from approximated types.

Reply:
Not done, for two reasons:
- from the linked documentation, intervals come in two flavours - year+month intervals, and day+hour+minute+second intervals. These two flavours can't be mixed in a single column, so using this type will probably lead to more problems than approximating as text.
- not sure, but I think the documentation is saying the Microsoft's ODBC driver supports intervals. SQL Server doesn't actually seem to have an interval type, or at least, not one I have found.
See https://docs.microsoft.com/en-us/sql/t-sql/data-types/data-types-transact-sql?view=sql-server-ver15

Reply:
in general it would be more useful to represent intervals as a double in seconds than anything text-based.

Reply:
This is not quite true of intervals - unfortunately there is no formula for converting a month into seconds. The interval type, in the working copies we have the support it (so far only PostGIS) can't be collapsed into any single number.

I also don't know if the user wants to do date-math, or to write conversion code - if I gave the user something machine-readable like number of seconds, then an advanced user could easily write a function that converted that into the datatype of their choice for display and manipulation. But all I know about this user is that they are checking out some data that they presumably didn't author, into a working copy type that the data is not native to, so perhaps they are outside the main use-case for this data. They might not have any extra tooling, or any developers to help them write it, but they might instead be glad of the ability to edit or at least view the intervals in a more-or-less human readable way.

Reply:
>  there is no formula for converting a month into seconds

ah, yes. ick.

Go with what you have already. 👍